### PR TITLE
General tag changes

### PR DESCRIPTION
**Pull request by zhamza99**
### What does this PR do?

Adds 'component' and 'span.kind' tags to several integrations to meet opentelemetry standards. 
Modifies tests to check for newly added tags per integration
Component is simply the library/integration name the span originated from.
Span.kind is better explained [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind) where it mentions the different types

Changes (string) system.pid to (float64) process_id as per new tracer requirements

Changes system tests to include testing for newly added 'component' and 'span.kind' tags

### Motivation

The tracer should meet the opentelemetry standard as more people adapt it and we want all information that users expect to be available for as many instances as possible

### Describe how to test/QA your changes

Tests for relevant libraries have been updated and passed. As long as the resulting spans have all values that are expected as described in the aforementioned link, things should be good.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

File: contrib/gorilla/mux/mux.go
Comment (dianashevchenko) [inline contrib/gorilla/mux/mux.go:10]: We sort import in a specific order, please put this alongside other gopkg imports.

File: contrib/99designs/gqlgen/tracer_test.go
Comment (dianashevchenko) [inline contrib/99designs/gqlgen/tracer_test.go:145]: It looks like you added Component and SpanKind to the ddtrace/ext/tags.go, but you don't use it everywhere. Please replace those string literals in libraries with a constant

File: contrib/go-redis/redis.v8/redis.go
Comment (dianashevchenko) [inline contrib/go-redis/redis.v8/redis.go:148]: redis.v7 / redis.v8 integrations have the same component name. I'm not sure, but it's worth looking up if it should instead be:
```suggestion
		tracer.Tag(ext.Component, "go-redis.v8"),
```

Comment (dianashevchenko) [reply contrib/go-redis/redis.v8/redis.go:148]: And .v7 in the other part as well

File: contrib/go.mongodb.org/mongo-driver/mongo/mongo.go
Comment (dianashevchenko) [inline contrib/go.mongodb.org/mongo-driver/mongo/mongo.go:52]: Span kind is limited to a few values as well: server, client, producer, consumer, internal. Maybe it's worth adding corresponding internal constants, in case we ever have to go back to this issue

**Review by dianashevchenko — CHANGES_REQUESTED** [review]
I will do more in-depth review later, with each integration.

File: contrib/go.mongodb.org/mongo-driver/mongo/mongo.go
Comment (zhamza99) [reply contrib/go.mongodb.org/mongo-driver/mongo/mongo.go:52]: Added

File: contrib/99designs/gqlgen/tracer_test.go
Comment (zhamza99) [reply contrib/99designs/gqlgen/tracer_test.go:145]: I disagree, I purposely left them out, because if someone were to change them in the future, this would force them to check tests to ensure that their change doesn't break anything. Otherwise people could change them without doing thorough checks of everything affected. This is only for tests that people often overlook

File: contrib/go-redis/redis.v8/redis.go
Comment (zhamza99) [reply contrib/go-redis/redis.v8/redis.go:148]: Done

File: contrib/99designs/gqlgen/tracer_test.go
Comment (dianashevchenko) [reply contrib/99designs/gqlgen/tracer_test.go:145]: If you check the tests though, they do use other constants

Comment (zhamza99) [reply contrib/99designs/gqlgen/tracer_test.go:145]: Implemented

**Review by zhamza99 — COMMENTED** [review]
All changes are done, waiting for review

File: ddtrace/tracer/tracer.go
Comment (knusbaum) [inline ddtrace/tracer/tracer.go:458]: We could use some information about why this is changing, as it doesn't seem to be related to the broader opentelemetry-related tag stuff.

* Why are we now setting pid/language tags on all spans?
* Why is the pid now a metric rather than metadata?
* Why is pid changing from `system.pid` to `process_id`?

File: ddtrace/ext/span_kind.go
Comment (knusbaum) [inline ddtrace/ext/span_kind.go:9]: > falls under the values of client, server, producer, consumer, and internal

I'm not sure I understand what this means. Can we possibly clarify slightly and link to the otel standard?

Also, Nit: let's get rid of the newline between this comment and const.

**Review by knusbaum — CHANGES_REQUESTED** [review]
It would be good to standardize the format of the Component a little more.

Currently, some of them are just the last part of the package:
`"gqlgen"` -> `contrib/99designs/gqlgen`

Some of them include path components:
`"Shopify/sarama"` -> `contrib/Shopify/sarama`

And some of them contain parts that aren't in the package path at all:
`"google-cloud-go/pubsub"` -> `contrib/cloud.google.com/go/pubsub.v1`


I'm thinking the full package name after the `contrib/` part might be best.
Open to other ideas.

File: ddtrace/ext/span_kind.go
Comment (zhamza99) [reply ddtrace/ext/span_kind.go:9]: This is the [link](https://github.com/open-telemetry/opentelemetry-specification/blob/9fa7c656b26647b27e485a6af7e38dc716eba98a/specification/trace/api.md#spankind) to the otel documents explaining what span kind is and the possible values. This .go file should explain the different values possible for span.kind, but if there are more changes necessary, I can add those

File: ddtrace/tracer/tracer.go
Comment (zhamza99) [reply ddtrace/tracer/tracer.go:458]: process_id needs to be a numeric value, and since span.setMeta only sets string values, we have to use span.setMetric to set the process_id as a numeric value

Comment (ahmed-mez) [reply ddtrace/tracer/tracer.go:458]: I'm definitely lacking context here, just want to make sure these two questions aren't about breaking changes. Also would this require a change in the trace-agent? Should we keep both fields instead of replacing?

> Why is the pid now a metric rather than metadata?
Why is pid changing from system.pid to process_id?

Would you confirm @knusbaum ?

Comment (zhamza99) [reply ddtrace/tracer/tracer.go:458]: Another detail about the system.pid, even though we are just renaming it with this PR, the backend has already been remapping and duplicating the value "system.pid" to "process.id" so it wouldn't produce any breaking changes. Renaming the tag and changing its value is fine as other tracers already have numerical process_id tags.

**Comment by zhamza99** [comment]
Implemented 'component' change to all libraries as well as adding changes to contrib/README.md to ensure further contributions implement the mandatory 'span.kind' and 'component' tags

**Review by ahmed-mez — DISMISSED** [review]
Thank you @zhamza99 ! Let's get a +1 from @knusbaum as well before merging.

File: contrib/aws/aws-sdk-go-v2/aws/aws.go
Comment (knusbaum) [inline contrib/aws/aws-sdk-go-v2/aws/aws.go:81]: This should be `"aws/aws-sdk-go-v2/aws"`

File: contrib/globalsign/mgo/mgo_test.go
Comment (knusbaum) [inline contrib/globalsign/mgo/mgo_test.go:139]: Nit: empty line

File: contrib/globalsign/mgo/mgo_test.go
Comment (knusbaum) [inline contrib/globalsign/mgo/mgo_test.go:213]: Nit: empty line

File: contrib/globalsign/mgo/mgo_test.go
Comment (knusbaum) [inline contrib/globalsign/mgo/mgo_test.go:256]: Nit: empty line

File: contrib/google.golang.org/grpc/grpc_test.go
Comment (knusbaum) [inline contrib/google.golang.org/grpc/grpc_test.go:149]: This looks like leftovers.

File: contrib/hashicorp/vault/vault_test.go
Comment (knusbaum) [inline contrib/hashicorp/vault/vault_test.go:139]: Nit: empty line

File: contrib/julienschmidt/httprouter/httprouter_test.go
Comment (knusbaum) [inline contrib/julienschmidt/httprouter/httprouter_test.go:49]: Nit: empty line

File: ddtrace/tracer/tracer.go
Comment (knusbaum) [reply ddtrace/tracer/tracer.go:458]: This doesn't answer the numeric change. Are we sure this needs to be a metric and not metadata? It was previously metadata. Fields like this that are used to connect things sometimes are very particular. 

I also still have concerns (put in slack) about this change, as it doesn't seem to match the spec, and I'm hoping we won't need to do this at all.

**Review by knusbaum — CHANGES_REQUESTED** [review]
Some minor nits/cleanup + remaining questions about pid/language.

**Comment by zhamza99** [comment]
> This doesn't answer the numeric change. Are we sure this needs to be a metric and not metadata? It was previously metadata. Fields like this that are used to connect things sometimes are very particular.

Process_id needs to be a numeric value as per new specifications. The meta tags of a span are stored in a string-to-string map which means that numerical values cannot be put into it. Instead, the metric tags of a span are in a string-to-numeric map which is where other numerical values are stored. Thus we use `span.setMetric` to set process_id instead of `span.setMeta` used before

In terms of the backend, other tracers already sent process_id through the metric tags so there isn't an issue there. In fact, the go tracer is the only one sending process_id as a string via the meta tags which is why to standardize everything across tracers, the value is being changed to a numeric value sent via the metric tags

**Review by knusbaum — APPROVED** [review]
> In terms of the backend, other tracers already sent process_id through the metric tags so there isn't an issue there. In fact, the go tracer is the only one sending process_id as a string via the meta tags which is why to standardize everything across tracers, the value is being changed to a numeric value sent via the metric tags

This was what I was looking for. Thanks for clarifying. This LGTM now.

**Review by ahmed-mez — APPROVED** [review]

**Pull request merged**
